concord-server/server/src/main/java/com/rabidgremlin/concord/resources/PhrasesResource.java

```diff
@@ -121,8 +122,7 @@ public Response uploadCsv(@ApiParam(hidden = true) @Auth Caller caller, List<Unl
 		   String phraseId = DigestUtils.md5Hex(unlabelledPhrase.getText());
 		   
 		   // remove any existing votes in case we are reloading data
-		   votesDao.deleteAllVotesForPhrase(phraseId);
-		   phrasesDao.upsert(phraseId, unlabelledPhrase.getText(), false);
+		   uploadDao.deleteExistingVotesAndUploadPhrase(phraseId, unlabelledPhrase);

```

Comment:
@rabidgremlin Do you think we should make the entire loop a transaction? Or if we fail keep the ones we managed to upload to the database.

Reply:
I think it should be all or nothing. So rollback if a single failure.